pan3d/dataset_builder.py

```diff
@@ -0,0 +1,284 @@
+import json
+import os
+import pyvista
+import xarray
+from pvxarray.vtk_source import PyVistaXarraySource
+from pyvista.trame.ui import plotter_ui
+
+from trame.decorators import TrameApp, change
+from trame.ui.vuetify import SinglePageWithDrawerLayout  # TODO: upgrade to vuetify 3
+from trame.app import get_server
+
+from pan3d.ui import AxisSelection, MainDrawer, RenderArea, Toolbar
+from pan3d.utils import initial_state, run_singleton_task
+
+
+@TrameApp()
+class DatasetBuilder:
+    def __init__(self, server=None, dataset_path=None, state=None):
+        if server is None:
+            server = get_server()
+
+        if isinstance(server, str):
+            server = get_server(server)
+
+        self.server = server
+        self.state.update(initial_state)
+        self.algorithm = PyVistaXarraySource()
+        self.plotter = pyvista.Plotter(off_screen=True, notebook=False)
+        self.plotter.set_background("lightgrey")
+        self.dataset = None
+        self.dataset_path = None
+        self.mesh = None
+        self.actor = None
+
+        self.ctrl.get_plotter = lambda: self.plotter
+        self.ctrl.reset = self.reset
+
+        if dataset_path:
+            self.set_dataset_path(dataset_path=dataset_path)
+        if state:
+            self.state.update(state)
+
+        # Fix version of vue
+        server.client_type = "vue2"  # TODO: upgrade to vue3
+
+    # -----------------------------------------------------
+    # Properties
+    # -----------------------------------------------------
+
+    @property
+    def state(self):
+        return self.server.state
+
+    @property
+    def ctrl(self):
+        return self.server.controller
+
+    @property
+    def viewer(self):
+        if self.layout is None:
+            # Build UI
+            self.layout = SinglePageWithDrawerLayout(self.server)
+            self.layout.title.set_text("Pan3D Viewer")
+            self.layout.footer.hide()
+            with self.layout:
+                with self.layout.toolbar:
+                    self.layout.toolbar.dense = True
+                    self.layout.toolbar.align = "center"
+                    Toolbar(reset=self.ctrl.reset)
+                with self.layout.drawer:
+                    MainDrawer()
+                with self.layout.content:
+                    plot_view = plotter_ui(
+                        self.ctrl.get_plotter(),
+                        interactive_ratio=1,
+                    )
+                    self.ctrl.view_update = plot_view.update
+                    self.ctrl.reset_camera = plot_view.reset_camera
+                    RenderArea(plot_view=plot_view)
+                AxisSelection()
+        return self.layout
```

Comment:
Maybe we should make layout internal by using `self._layout`

Reply:
Ok changed in [75c2ec7](https://github.com/Kitware/pan3d/pull/28/commits/75c2ec71f08e87d7a576740f3bc55b067d1904cb)